mono/metadata/appdomain.c

```diff
@@ -379,27 +384,29 @@ mono_get_corlib_version (void)
 const char*
 mono_check_corlib_version (void)
 {
-	const char* res;
+	char *result = NULL;
 	MONO_ENTER_GC_UNSAFE;
-	res = mono_check_corlib_version_internal ();
-	MONO_EXIT_GC_UNSAFE;
-	return res;
-}
-
-static const char *
-mono_check_corlib_version_internal (void)
-{
-	int version = mono_get_corlib_version ();
-	if (version != MONO_CORLIB_VERSION)
-		return g_strdup_printf ("expected corlib version %d, found %d.", MONO_CORLIB_VERSION, version);
+	char *version = mono_get_corlib_version ();
+	if (!version) {
+		result = g_strdup_printf ("expected corlib string (%s) but not found or not string", MONO_CORLIB_VERSION);
+		goto exit;
+	}
+	if (strcmp (version, MONO_CORLIB_VERSION) != 0) {
+		result = g_strdup_printf ("expected corlib version %s, found %s.", MONO_CORLIB_VERSION, version);
```

Comment:
Let's improve the error message while we're here, something like: "The runtime didn't find the mscorlib.dll it expected. Expected interface version %s but found %s. Check that your runtime and class libraries are matching."

Reply:
Agreed.